pkg/controller/controllerutil/controllerutil.go

```diff
@@ -168,3 +169,48 @@ func mutate(f MutateFn, key client.ObjectKey, obj runtime.Object) error {
 
 // MutateFn is a function which mutates the existing object into it's desired state.
 type MutateFn func() error
+
+// AddFinalizer accepts a metav1 object and adds the provided finalizer if not present.
+func AddFinalizer(o metav1.Object, finalizer string) {
+	f := o.GetFinalizers()
+	for _, e := range f {
+		if e == finalizer {
+			return
+		}
+	}
+	o.SetFinalizers(append(f, finalizer))
+}
+
+// AddFinalizerIfPossible tries to convert a runtime object to a metav1 object and add the provided finalizer.
+// It returns an error if the provided object cannot provide an accessor.
+func AddFinalizerIfPossible(o runtime.Object, finalizer string) error {
+	m, err := meta.Accessor(o)
+	if err != nil {
+		return err
+	}
+	AddFinalizer(m, finalizer)
+	return nil
+}
+
+// RemoveFinalizer accepts a metav1 object and removes the provided finalizer if present.
+func RemoveFinalizer(o metav1.Object, finalizer string) {
+	f := o.GetFinalizers()
+	for i, e := range f {
```

Comment:
Nit: This is nicer with `k8s.io/apimachinery/pkg/util/sets.String`

Reply:
found this today coincidentally. can swap it out 👍

Reply:
better?